### ISSUE TITLE
Overlay ctm, nothing will overlay on grass

### ISSUE THREAD

**Issue by MisterQuiggly** [sp614x/optifine#562]
I've tried multiple times but I cannot get a texture to overlay on grass, but I can get grass to overlay on other blocks.
Right now I'm trying to get sand to overlay on grass. It will overlay on other blocks but not grass.
What I have in the properties file:

> matchTiles=grass_top grass
> method=overlay
> faces=top sides
> tiles=0-16
> connectTiles=sand

I use that same set up for everything, but it just doesn't work when I want a block to connect to grass.

**Comment by Filmjolk** [comment]
Do you have a repeat pattern or other ctm on your grass? If you do, overlay does not work. I use model variants to get random tiles just to be sure, although random should work with overlay, I just don;t like to take any risks.

**Comment by Brottweiler** [comment]
Try replace `Tiles` with `Blocks`, so `matchTiles` with `matchBlocks` and `connectTiles` with `connectBlocks`. That's how I have it in my file.

**Comment by MisterQuiggly** [comment]
@Filmjolk I do not have any randomized grass or sand texture, nor any other method. 

@Brottweiler I tried switching it with Blocks and it also did not work. Something strange I find about this is that it doesn't work on two computers that I tried, and also that its only grass that blocks won't 'bleed' onto. I got in contact with someone on the Minecraft forums who had successfully gotten a block to overlay onto grass. I asked him if he could send over the files for that one block specifically, I applied it to my pack, and it didn't work. I then asked if he could send his entire pack over, I tried it, and it still did not work. This just baffled me and it made absolutely no sense. I also asked if he could send over his optionsof.txt and tried applying it so see if it worked, but it didn't.

**Comment by Filmjolk** [comment]
@MisterQuiggly Please give us a link to your pack. It's hard to determine what's wrong without seeing it. I have spent a lot of time playing around with Overlay recently on several packs. I should be able to determine what the cause is (if it's the pack itself and not your client).

**Comment by MisterQuiggly** [comment]
@Filmjolk Here is the link. https://www.dropbox.com/s/shuapsrhf3uf3s3/My%20Default%20Edit.zip?dl=0
I've tried to look for everything, I don't know why it would only be the grass that has a problem. In the pack you may notice I have a grass and sand CTM. Those have been disabled, and even before they were added it didn't work.

**Comment by Filmjolk** [comment]
@MisterQuiggly I went over it just now and like I suspected, you have 2 different repeat patterns for sand for different biomes. They are located in the nature folder. Remove repeat patterns or overlay wont work.

**Comment by MisterQuiggly** [comment]
@Filmjolk Yeah I know, but like I said I had it disabled. And it wasn't working before I added the sand as well.

**Comment by Filmjolk** [comment]
@MisterQuiggly  Have you tried a standalone pack with only the grass overlay and nothing else? Because the overlay works for me. Sounds like you have incompatible gpu/corrupted optifine or minecraft if you can't make it work with only the overlay by itself.

**Comment by MisterQuiggly** [comment]
@Filmjolk Made a new one with only sand overlaying on grass, didn't work.

**Comment by Filmjolk** [comment]
@MisterQuiggly  try to turn on your mipmap levels, set it to 1-4 not 0

**Comment by MisterQuiggly** [comment]
@Filmjolk Didn't work

**Comment by Filmjolk** [comment]
well, if you get it working on one computer, try to duplicate all the settings onto the other. It seems that various settings in optifine can cause grass overlay? or the tint to stack over the ctm overlay.

**Comment by MisterQuiggly** [comment]
@Filmjolk I haven't tried this yet because I haven't had a lot of time. But I just thought I would mention:

> Well, if you get it working on one computer, try to duplicate all the settings onto the other.

This didn't work on either of the two computers I tried. And like mentioned above, it was only the grass. I will try with your settings. I have tried minecraft:grass.

> Here are my settings which currently make grass overlay work for me:

Are you talking about having grass overlay onto other blocks, or blocks overlay onto grass?

> I would try to remove the overlay and tintindex on the grass model and see if it allows the ctm overlay

With this, are you talking about editing in the json file for the model?

**Comment by sp614x** [comment]
Minecraft renders the blocks in several layers in a fixed orded (solid, cutout_mipped, cutout and translucent). 
Grass is rendered in layer "cutout_mipped" because of the biome colored sides. 
To render an overlay on top of the grass it has to be rendered in layer "cutout_mipped" with properties file coming after the grass properties or in layer "translucent".

It should be possible to configure the sand overlay to use layer "cutout_mipped" and name the properties file so that it comes after the grass overlay: https://github.com/sp614x/optifine/blob/master/OptiFineDoc/doc/ctm.properties#L189
This may decrease the FPS when many sand blocks are rendered (in a desert) as the "cutout" layer is slower than the "solid" layer.

**Comment by MisterQuiggly** [comment]
@sp614x Many things happened when experimenting. Currently, I have two blocks trying to overlay onto grass; cobblestone and sand. After you sent that message I didn't touch the cobblestone. I did add "layer=cutout_mipped" to the sand texture. I turned on mipmaps and reloaded the resourcepack. The sand did not change, this did not work. But, without editing the cobblestone file, the cobblestone did overlay onto the grass when I turned on mipmap. So I went into the cobblestone texture and added "layer=cutout" and reloaded the resourcepack and turned off mipmaps. So now after this, the cobblestone overlays onto grass no matter what, which is good. Sand, on the other hand, will still not overlay on the grass. I've tried cutout_mipped, cutout, and translucent as well as turning mipmap on for the ones that said it needed to be on. So after all of this, the sand will still not overlay on the grass. I know the sand ctm is working because its overlay on another block I had it set to.

**Comment by sp614x** [comment]
Check if the properties file names are in the correct alphabetical order. You can also prefix them with a number to make it easier, for example:
1_grass.properties
2_sand.properties
3_cobblestone.properties

**Comment by MisterQuiggly** [comment]
@sp614x Are you saying that they should all be in the same folder? What I have currently is: mcpatcher/ctm/overlays/*.
the asterisk would represent which overlay it is. (Example, grass_overlay, cobblestone_overlay, sand_overlay) I could compile them all into one folder if that is what you are referencing?

**Comment by sp614x** [comment]
Then the folders should be named alphabetically in the order in which they have to be applied, for example:
1_grass
2_sand
3_cobblestone

**Comment by MisterQuiggly** [comment]
@sp614x I renamed them to 1_grass, 2_sand, and 3_cobblestone and the sand still does not work, cobblestone still works though.

**Comment by sp614x** [comment]
Resource pack link?

**Comment by MisterQuiggly** [comment]
@sp614x Most recent version: https://www.dropbox.com/s/shuapsrhf3uf3s3/My%20Default%20Edit.zip?dl=0

**Comment by StereoMadnessss** [comment]
same issue with biome dependent textures
![2022-03-20_20 44 45](https://user-images.githubusercontent.com/77188051/159202653-9b0924b3-ec08-423e-9429-4b1b34de6806.png)
CTM file:

<img width="228" alt="Screen Shot 2022-03-20 at 8 45 36 PM" src="https://user-images.githubusercontent.com/77188051/159202719-50c357dd-6398-4cae-a2c7-320a7ccbb23a.png">

**Comment by MisterQuiggly** [comment]
> same issue with biome dependent textures ![2022-03-20_20 44 45](https://user-images.githubusercontent.com/77188051/159202653-9b0924b3-ec08-423e-9429-4b1b34de6806.png) CTM file:
> 
> <img alt="Screen Shot 2022-03-20 at 8 45 36 PM" width="228" src="https://user-images.githubusercontent.com/77188051/159202719-50c357dd-6398-4cae-a2c7-320a7ccbb23a.png">

Don't you need something like seen here? I don't really remember off of the top of my head lol.
https://github.com/sp614x/optifine/blob/42714a802ce03b5786c301f54aa924643eb6057b/OptiFineDoc/doc/ctm.properties#L148-L156

**Comment by MrRedstoneA1** [comment]
I was having issues with nothing overlaying on grass until I discovered it was caused by the better foliage mod's "connected grass" setting. If that's turned on, nothing will overlay on full side grass blocks. I'm posting this in case anyone else is having issues.
